Zero Network Milestone 1

ashWhiteHat opened the pull request:
# Milestone Delivery Checklist

- [x] The [milestone-delivery-template.md](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/deliveries/milestone-delivery-template.md) has been copied and updated.
- [ ] The [invoice form :pencil:](https://forms.gle/LSRr7PCjBpEbKGh89) has been filled out for this milestone.
- [ ] This pull request is being made by the same account as the accepted application.
- [ ] In case of acceptance, the payment will be transferred to the BTC/ETH payment address provided in the application.
- [ ] The delivery is according to the [Guidelines for Milestone Deliverables](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/milestone-deliverables-guidelines.md).

Link to the application pull request: https://github.com/w3f/Grants-Program/pull/909

keeganquigley commented:
Hi @ashWhiteHat thank you for the delivery. Someone from the grants team will look at it shortly.

dsm-w3f commented:
@ashWhiteHat thank you for the milestone submission. Please see the [evaluation documents](https://github.com/w3f/Grant-Milestone-Delivery/blob/master/evaluations/zero-network_1_dsm-w3f.md) and provide proper answers and fixes. After that, let me know to continue with the evaluation.

ashWhiteHat commented:
Hi @dsm-w3f 
Thank you for the polite review.

I am going to create example which uses node and client rpc, and add tests.
Some of issues were solved.

> Would be better to use HTTPS instead.

I fixed it with https://github.com/zero-network/zero/pull/62/commits/5710b970fe220fd83ff79c18d9647aef8bf82005
> There are some typos there, such as "toools" and "Summerize"

I fixed typos with https://github.com/zero-network/zero/pull/62/commits/2c834e1f1a548b5168ebeff8638d8cc003baaab9
> could need a more careful security assessment in the future

Exactly, before we connect with Kusama, we are going to do that.
> Please, add some examples ready for execution so I continue with the evaluation
we expect docker compose file to spin up the system for manual testing using a substrate node

We create [repo](https://github.com/zero-network/node) and make test interacting through rpc.
Please wait for a while.
> The coverage could be improved

I added ci and calculated coverage.
I am going to improve.

- Pallets

| Filename | Regions | Missed Regions | Cover | Functions | Missed Functions | Executed | Lines | Missed Lines | Cover |
|---|---|---|---|---|---|---|---|---|---|
|confidential_transfer/circuit     | 34 | 10 | 70.59% | 11 |  4 | 63.64% | 145 |  4 |  97.24% |
|confidential_transfer/lib         | 33 | 15 | 54.55% | 11 |  5 | 54.55% |  49 | 13 |  73.47% |
|confidential_transfer/mock        | 10 |  2 | 80.00% |  4 |  0 |100.00% |  72 |  0 | 100.00% |
|confidential_transfer/tests       | 20 |  0 |100.00% | 10 |  0 |100.00% |  65 |  0 | 100.00% |
|encrypted_balance/lib             |110 | 32 | 70.91% | 45 | 13 | 71.11% | 176 | 25 |  85.80% |
|encrypted_balance/tests           |  4 |  0 |100.00% |  2 |  0 |100.00% |  19 |  0 | 100.00% |
|encrypted_balance/tests_composite | 10 |  2 | 80.00% |  6 |  1 | 83.33% |  40 |  6 |  85.00% |
|encrypted_balance/tests_local     | 19 |  2 | 89.47% | 11 |  1 | 90.91% |  74 |  6 |  91.89% |
|encrypted_balance/weights         |  8 |  8 |  0.00% |  8 |  8 |  0.00% |  40 | 40 |   0.00% |
|plonk/lib                         | 40 | 22 | 45.00% | 13 |  7 | 46.15% |  67 | 27 |  59.70% |
|plonk/mock                        | 13 |  3 | 76.92% |  5 |  0 |100.00% |  36 |  0 | 100.00% |
|plonk/tests                       | 16 |  1 | 93.75% |  8 |  0 |100.00% |  58 |  0 | 100.00% |
|plonk/types                       |  5 |  5 |  0.00% |  5 |  5 |  0.00% |   7 |  7 |   0.00% |

- Primitives

| Filename | Regions | Missed Regions | Cover | Functions | Missed Functions | Executed | Lines | Missed Lines | Cover |
|---|---|---|---|---|---|---|---|---|---|
|bls12_381/fq                                                    | 23 | 15 | 34.78% |  14 |  7  | 50.00% | 131 |  72  |  45.04% |
|bls12_381/fqn                                                   | 66 | 21 | 68.18% |  30 | 10  | 66.67% | 196 |  93  |  52.55% |
|bls12_381/fr                                                    | 75 | 17 | 77.33% |  34 |  5  | 85.29% | 228 |  34  |  85.09% |
|bls12_381/g1                                                    |118 | 48 | 59.32% |  43 | 19  | 55.81% | 354 | 179  |  49.44% |
|bls12_381/g2                                                    | 65 | 51 | 21.54% |  34 | 23  | 32.35% | 233 | 222  |   4.72% |
|bls12_381/gt                                                    | 23 | 20 | 13.04% |  15 | 12  | 20.00% |  51 |  42  |  17.65% |
|bls12_381/lib                                                   |  1 |  0 |100.00% |   1 |  0  |100.00% |   1 |   0  | 100.00% |
|crypto/arithmetic/bits_256/coordinate/projective                | 25 | 25 |  0.00% |   4 |  4  |  0.00% |  59 |  59  |   0.00% |
|crypto/arithmetic/bits_256/limbs/normal                         | 26 |  2 | 92.31% |   9 |  0  |100.00% | 115 |   2  |  98.26% |
|crypto/arithmetic/bits_256/represent                            | 29 |  1 | 96.55% |  11 |  1  | 90.91% |  91 |   3  |  96.70% |
|crypto/arithmetic/bits_384/coordinate/projective                | 25 | 11 | 56.00% |   4 |  2  | 50.00% |  59 |  14  |  76.27% |
|crypto/arithmetic/bits_384/limbs/normal                         | 26 |  2 | 92.31% |   9 |  0  |100.00% | 169 |   2  |  98.82% |
|crypto/arithmetic/bits_384/represent                            | 29 | 15 | 48.28% |  11 |  5  | 54.55% | 105 |  36  |  65.71% |
|crypto/arithmetic/utils                                         | 13 |  4 | 69.23% |   7 |  2  | 71.43% |  22 |   3  |  86.36% |
|crypto/dress/curve                                              | 38 |  9 | 76.32% |  26 |  8  | 69.23% |  95 |  25  |  73.68% |
|crypto/dress/curve/group                                        | 85 | 26 | 69.41% |  31 | 16  | 48.39% | 153 |  62  |  59.48% |
|crypto/dress/curve/test                                         | 56 |  0 |100.00% |   9 |  0  |100.00% | 151 |   0  | 100.00% |
|crypto/dress/extension_field                                    | 13 |  4 | 69.23% |  11 |  4  | 63.64% |  27 |  10  |  62.96% |
|crypto/dress/extension_field/field                              |  2 |  2 |  0.00% |   2 |  2  |  0.00% |   8 |   8  |   0.00% |
|crypto/dress/extension_field/group                              | 30 |  1 | 96.67% |  13 |  1  | 92.31% |  60 |   3  |  95.00% |
|crypto/dress/extension_field/ring                               | 45 | 17 | 62.22% |   8 |  4  | 50.00% |  46 |  22  |  52.17% |
|crypto/dress/field                                              | 20 | 11 | 45.00% |  10 |  2  | 80.00% |  33 |   9  |  72.73% |
|crypto/dress/field/field                                        |  2 |  2 |  0.00% |   2 |  2  |  0.00% |   8 |   8  |   0.00% |
|crypto/dress/field/group                                        | 30 |  3 | 90.00% |  26 |  2  | 92.31% |  80 |   7  |  91.25% |
|crypto/dress/field/ring                                         | 45 | 15 | 66.67% |   8 |  3  | 62.50% |  46 |  18  |  60.87% |
|crypto/dress/field/test                                         | 31 |  1 | 96.77% |   6 |  0  |100.00% |  81 |   1  |  98.77% |
|crypto/dress/pairing/bls12/extension_field                      | 31 |  5 | 83.87% |  26 |  4  | 84.62% | 218 |  17  |  92.20% |
|crypto/dress/pairing/bls12/g2                                   | 16 |  1 | 93.75% |   4 |  0  |100.00% |  83 |   4  |  95.18% |
|crypto/dress/pairing/bls12/range_field                          | 18 |  2 | 88.89% |   7 |  2  | 71.43% | 291 | 134  |  53.95% |
|crypto/lib                                                      |  1 |  0 |100.00% |   1 |  0  |100.00% |   1 |   0  | 100.00% |
|elgamal/lib                                                     | 61 | 15 | 75.41% |  26 |  5  | 80.77% | 152 |  19  |  87.50% |
|jubjub/curve                                                    |140 | 86 | 38.57% | 104 | 66  | 36.54% | 605 | 269  |  55.54% |
|jubjub/error                                                    |  2 |  2 |  0.00% |   1 |  1  |  0.00% |   1 |   1  |   0.00% |
|jubjub/fp                                                       | 91 | 24 | 73.63% |  25 |  7  | 72.00% | 200 |  31  |  84.50% |
|jubjub/lib                                                      |  1 |  0 |100.00% |   1 |  0  |100.00% |   1 |   0  | 100.00% |
|kzg/commitment                                                  | 12 |  0 |100.00% |   6 |  0  |100.00% |  33 |   0  | 100.00% |
|kzg/fft                                                         | 70 |  4 | 94.29% |  41 |  3  | 92.68% | 256 |   3  |  98.83% |
|kzg/keypair                                                     | 10 |  3 | 70.00% |   7 |  3  | 57.14% |  52 |  25  |  51.92% |
|kzg/lib                                                         |  1 |  0 |100.00% |   1 |  0  |100.00% |   1 |   0  | 100.00% |
|kzg/poly                                                        | 62 | 32 | 48.39% |  34 | 13  | 61.76% | 157 |  71  |  54.78% |
|kzg/witness                                                     |  1 |  1 |  0.00% |   1 |  1  |  0.00% |   3 |   3  |   0.00% |
|pairing/lib                                                     | 46 |  2 | 95.65% |   7 |  0  |100.00% |  61 |   2  |  96.72% |
|plonk/bit_iterator                                              | 13 |  2 | 84.62% |   6 |  2  | 66.67% |  32 |   2  |  93.75% |
|plonk/commitment_scheme/kzg10/commitment                        | 10 |  7 | 30.00% |   9 |  6  | 33.33% |  18 |   9  |  50.00% |
|plonk/commitment_scheme/kzg10/key                               |144 | 43 | 70.14% |  33 |  7  | 78.79% | 308 |  32  |  89.61% |
|plonk/commitment_scheme/kzg10/proof                             | 13 |  6 | 53.85% |  10 |  5  | 50.00% |  61 |   5  |  91.80% |
|plonk/commitment_scheme/kzg10/srs                               | 27 |  8 | 70.37% |  10 |  4  | 60.00% |  71 |  10  |  85.92% |
|plonk/composer                                                  |130 | 16 | 87.69% |  40 |  5  | 87.50% | 720 |  54  |  92.50% |
|plonk/composer/builder                                          | 18 |  2 | 88.89% |  13 |  2  | 84.62% | 110 |   2  |  98.18% |
|plonk/composer/compiler                                         | 33 |  8 | 75.76% |   4 |  0  |100.00% | 380 |   0  | 100.00% |
|plonk/composer/polynomial                                       |  4 |  4 |  0.00% |   3 |  3  |  0.00% |   3 |   3  |   0.00% |
|plonk/composer/prover                                           | 60 | 14 | 76.67% |   7 |  2  | 71.43% | 330 |   4  |  98.79% |
|plonk/composer/verifier                                         |  7 |  1 | 85.71% |   3 |  0  |100.00% |  52 |   4  |  92.31% |
|plonk/constraint_system/constraint                              | 38 | 13 | 65.79% |  35 | 10  | 71.43% | 119 |  12  |  89.92% |
|plonk/constraint_system/ecc                                     |  7 |  4 | 42.86% |   7 |  4  | 42.86% |  13 |   4  |  69.23% |
|plonk/constraint_system/witness                                 | 11 |  4 | 63.64% |   9 |  3  | 66.67% |  13 |   3  |  76.92% |
|plonk/debugger                                                  | 57 | 28 | 50.88% |  21 | 10  | 52.38% | 186 | 123  |  33.87% |
|plonk/error                                                     | 26 | 26 |  0.00% |   4 |  4  |  0.00% |  46 |  46  |   0.00% |
|plonk/fft/domain                                                | 31 | 11 | 64.52% |   9 |  2  | 77.78% |  92 |  15  |  83.70% |
|plonk/fft/evaluations                                           | 29 | 26 | 10.34% |  16 | 13  | 18.75% |  59 |  49  |  16.95% |
|plonk/fft/polynomial                                            |164 | 78 | 52.44% |  37 | 11  | 70.27% | 269 |  94  |  65.06% |
|plonk/lib                                                       |  1 |  0 |100.00% |   1 |  0  |100.00% |   1 |   0  | 100.00% |
|plonk/permutation                                               |265 |  2 | 99.25% |  69 |  1  | 98.55% | 986 |   2  |  99.80% |
|plonk/proof_system/linearization_poly                           | 10 |  8 | 20.00% |   8 |  6  | 25.00% | 200 |   6  |  97.00% |
|plonk/proof_system/proof                                        | 27 | 10 | 62.96% |  15 |  6  | 60.00% | 365 |   7  |  98.08% |
|plonk/proof_system/quotient_poly                                | 14 |  0 |100.00% |   7 |  0  |100.00% | 246 |   0  | 100.00% |
|plonk/proof_system/widget                                       | 10 |  7 | 30.00% |   8 |  5  | 37.50% |  36 |   5  |  86.11% |
|plonk/proof_system/widget/arithmetic/proverkey                  |  6 |  4 | 33.33% |   5 |  3  | 40.00% |  73 |   3  |  95.89% |
|plonk/proof_system/widget/arithmetic/verifierkey                |  5 |  4 | 20.00% |   4 |  3  | 25.00% |  30 |   3  |  90.00% |
|plonk/proof_system/widget/ecc/curve_addition/proverkey          |  5 |  3 | 40.00% |   5 |  3  | 40.00% |  96 |   3  |  96.88% |
|plonk/proof_system/widget/ecc/curve_addition/verifierkey        |  4 |  3 | 25.00% |   4 |  3  | 25.00% |  44 |   3  |  93.18% |
|plonk/proof_system/widget/ecc/scalar_mul/fixed_base/proverkey   |  8 |  4 | 50.00% |   7 |  3  | 57.14% | 143 |   3  |  97.90% |
|plonk/proof_system/widget/ecc/scalar_mul/fixed_base/verifierkey |  5 |  4 | 20.00% |   4 |  3  | 25.00% |  59 |   3  |  94.92% |
|plonk/proof_system/widget/logic/proverkey                       |  8 |  4 | 50.00% |   7 |  3  | 57.14% | 109 |   3  |  97.25% |
|plonk/proof_system/widget/logic/verifierkey                     |  5 |  4 | 20.00% |   4 |  3  | 25.00% |  35 |   3  |  91.43% |
|plonk/proof_system/widget/permutation/proverkey                 | 12 |  4 | 66.67% |  11 |  3  | 72.73% | 216 |   3  |  98.61% |
|plonk/proof_system/widget/permutation/verifierkey               |  5 |  4 | 20.00% |   4 |  3  | 25.00% |  60 |   3  |  95.00% |
|plonk/proof_system/widget/range/proverkey                       |  6 |  3 | 50.00% |   6 |  3  | 50.00% |  63 |   3  |  95.24% |
|plonk/proof_system/widget/range/verifierkey                     |  4 |  3 | 25.00% |   4 |  3  | 25.00% |  26 |   3  |  88.46% |
|plonk/runtime                                                   |  7 |  5 | 28.57% |   6 |  4  | 33.33% |  14 |   4  |  71.43% |
|plonk/transcript                                                |  5 |  0 |100.00% |   5 |  0  |100.00% |  39 |   0  | 100.00% |
|plonk/util                                                      | 30 |  0 |100.00% |  11 |  0  |100.00% |  72 |   0  | 100.00% |
|src/lib                                                             | 1  |  0 |100.00% |   1 |  0  |100.00% |   1 |   0  | 100.00% |
|**TOTAL**                                                              |3091 |979|68.33%  |1281 | 434 |66.12%  |11361| 2172 |  80.88% |

> please review your code running regarding those errors and also the warnings from cargo clippy

I fixed errors with https://github.com/zero-network/zero/pull/62/commits/454768669c5753df56fb77cf2607373326f631ad.
Some warnings still are not resolved.
I am going to work on that.

## Todo
- [ ] Test using node and client through RPC
- [ ] Improve code coverage
- [ ] Fix warnings
